### PR TITLE
Feature/websocket client

### PR DESCRIPTION
**Pull request by slaff**
Rebase of #258

### CODE REVIEW

**Comment by piperpilot** [comment]
Glad to see the websocket client resurrected.  What do we need to get this brought into 2.1.5?

**Comment by slaff** [comment]
> What do we need to get this brought into 2.1.5?

@piperpilot Test and report if it is working for you. The report should contain SDK (open, alt, etc) used, version (1.4, 1.5, etc) and OS.

**Comment by slaff** [comment]
Note to myself: improved Websocket Client from @avr39-ripe is in the works https://github.com/avr39-ripe/Sming/blob/feature/websockets-client/.

**Comment by piperpilot** [comment]
I will try to test this tomorrow and make sure it all works.

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+				debugf("Got Disconnect request from server.");
+				//RFC requires we return a close op code before closing the connection
+				disconnect();
+			}
+			else if (op == 0x09)
+			{
+				debugf("Got ping ...");
+				sendPong(); //Need to send Pong in response to Ping
+			}
+			else if (op = 0x10)
```
Comment (slaff) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: @hrsavla That's a typo. Should be `else if (op == 0x10)`

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+	 Opcodes
+	 0x00: this frame continues the payload from the last.
+	 0x01: this frame includes utf-8 text data.
+	 0x02: this frame includes binary data.
+	 0x08: this frame terminates the connection.
+	 0x09: this frame is a ping.
+	 0x0A: this frame is a pong.
+	 */
+	uint8_t buffer[500];
+	uint8_t i = 0;
```
Comment (ADiea) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: must be uint16_t

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+
+	 Opcodes
+	 0x00: this frame continues the payload from the last.
+	 0x01: this frame includes utf-8 text data.
+	 0x02: this frame includes binary data.
+	 0x08: this frame terminates the connection.
+	 0x09: this frame is a ping.
+	 0x0A: this frame is a pong.
+	 */
+	uint8_t buffer[500];
```
Comment (ADiea) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: on my branch I have put this on heap char* buffer = new char[10 + length];

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+{
+	if (buf == NULL)
+	{
+		// Disconnected, close it
+		TcpClient::onReceive(buf);
+	}
+	else
+	{
+		uint16_t size = buf->tot_len;
+		char* data = new char[size + 1];
```
Comment (ADiea) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: data should be checked for NULL

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+		data[size] = '\0';
+
+		//  debugf("%s", data); //print received buffer
+		switch (Mode)
+		{
+		case ws_Connecting:
+			if (verifyKey(data, size) == true)
+			{
+				Mode = ws_Connected;
+				this->connectedcallback(Mode);
```
Comment (ADiea) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: on my branch connectedcallback() is called after sendPing() I can't remember why

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+			}
+			else
+			{
+				debugf("Unknown opcode : %d ", op);
+				//Or not start of package if we failed to parse the entire previous one
+			}
+			break;
+		}
+		TcpClient::onReceive(buf);
+	}
```
Comment (ADiea) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: heap allocated data should be freed here, it was copied in pbuf

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+}
+
+/* Function Name: sendMessage
+ * Description: Send Text Stringmessage to Websocket Server
+ *              Max length of message permitted is 0xffff (65535) bytes
+ * Parameters: Str - String to be sent to websocket server
+ */
+void WebsocketClient::sendMessage(String str)
+{
+	uint16_t size = str.length() + 1;
```
Comment (ADiea) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: all function can be written as sendMessage(str.c_str(), str.length());

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+					uint8_t mask[4];
+					if (masked > 0)
+					{
+						mask[0] = data[cnt++];
+						mask[1] = data[cnt++];
+						mask[2] = data[cnt++];
+						mask[3] = data[cnt++];
+					}
+
+					uint8_t rxdata[len + 1]; //Max 16bit length message, so 65kbyte ...
```
Comment (ADiea) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: should be carefull here 65k on stack is huge will kill esp instantly, very suited for overflow attacks.
I would allocate on heap a max size and the application must be written in such a way to never send more than x bytes

**Comment by slaff** [comment]
@ADiea Thanks for the review.  Can you make PR for the recommended changes? The PR should be issued for this repo https://github.com/slaff/Sming/tree/feature/websocket-client

**Comment by avr39-ripe** [comment]
@ADiea thanks for hints! I rewrite this almost from scratch.. you can track my progress in https://github.com/avr39-ripe/Sming/tree/feature/websockets-client/Sming/SmingCore/Network 
I wish to make general WebsockeFrameClass to be used either by websocket server and client, no need to have two slightly different copy of code. Will be great to hear your comments/remarks when my work reaches some milestone.. 
@slaff I do not think that we should further correct original PR code.

**Comment by slaff** [comment]
> I wish to make general WebsockeFrameClass to be used either by websocket server and client

@avr39-ripe Then in the coming days I will close this PR without merging it and will wait for your PR with the WebsockeFrameClass.

**Comment by ADiea** [comment]
@slaff a litt;e difficult + cannot test because I have a file version with SSL included so it would not work directly I have to pull your branch first etc. 
I have made manual diff to comment on changes for this PR it's faster to edit the code directly in my opinion

@avr39-ripe thanks for effort! 
I am now focusing on my project and the websock client and server are now more or less mature and proved in several day tests
Now I have very little time for playing with websocks and wireshark - but I will try :) 
What is the status of your work is it working?

**Comment by avr39-ripe** [comment]
@slaff I think you'r right, close this for now.
@ADiea Not yet, just started to write |WebsocketFrameClass yesterday, but previouse implementation with some dirty hacks works nicely with binary websockets.. So when It will be ready for public testing I will kindly ask you for review and test.

**Comment by slaff** [comment]
See PR #809 for follow-up.

**Pull request closed without merging**